<a href="https://colab.research.google.com/github/hirenpatel1404/AI_lab/blob/main/Lesson_4_assignment_hiren.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Lesson 4 Assignment**
### **Data Processing - Python Development and Pandas**

**Copy each task (including'Your code here:') to a different cell and complete the codes for each task. Add  your own comments where necessary**

Remember to push your changes to github and submit a PDF copy of your completed notebook with the outputs.

In [61]:
# Python and Pandas Assessment Notebook

import pandas as pd
import numpy as np
from functools import reduce

# Task 1: Create a DataFrame using Pandas

In [62]:
## Build the data as a dictionary — keys become columns, lists become the data
data = {
    "Name": ["Alice", "Bob", "Charlie", "David"],
    "Age": [25, 30, 35, 28],
    "City": ["New York", "San Francisco", "Los Angeles", "Chicago"]
}



In [63]:
# Turn the dictionary into a DataFrame (a table)
df = pd.DataFrame(data)

In [64]:
# Show the table
print(df)

      Name  Age           City
0    Alice   25       New York
1      Bob   30  San Francisco
2  Charlie   35    Los Angeles
3    David   28        Chicago


In [65]:
# Check that data types are correct
print("\nData types:")
print(df.dtypes)


Data types:
Name    object
Age      int64
City    object
dtype: object


### Task 1 Interpretation
I created a table (DataFrame) with three columns: Name, Age, and City.
- Name and City are text (`object` type).
- Age is a whole number (`int64` type).
The data types are correct for the kind of data each column holds.

# Task 2: Row and Column Manipulation

In [66]:

# Drop the 'City' column from the DataFrame created in Task 1
df_no_city = df.drop("City", axis=1)
print(df_no_city)

      Name  Age
0    Alice   25
1      Bob   30
2  Charlie   35
3    David   28


In [67]:
# Confirm which columns are left
print("\nColumns remaining:", list(df_no_city.columns))


Columns remaining: ['Name', 'Age']


# Task 3: Handling Null Values

In [68]:
# Create a DataFrame with a null value (np.nan) in each column
df_nulls = pd.DataFrame({
    "Name": ["Alice", np.nan, "Charlie"],
    "Age": [25, 30, np.nan],
    "City": [np.nan, "San Francisco", "Los Angeles"]
})

In [69]:
# Show the table with nulls
print("Before filling nulls:")
print(df_nulls)

Before filling nulls:
      Name   Age           City
0    Alice  25.0            NaN
1      NaN  30.0  San Francisco
2  Charlie   NaN    Los Angeles


In [70]:
# Check how many nulls are in each column
print("\nNull count per column:")
print(df_nulls.isnull().sum())


Null count per column:
Name    1
Age     1
City    1
dtype: int64


In [71]:
# Fill nulls: 0 for the numeric Age column, "Unknown" for text columns
df_filled = df_nulls.fillna({"Name": "Unknown", "Age": 0, "City": "Unknown"})

In [72]:
#REPLACE AGE 0 WITH MEAN VALUE
df_filled["Age"] = df_filled["Age"].replace(0, df_filled["Age"].mean())
print(df_filled)

      Name        Age           City
0    Alice  25.000000        Unknown
1  Unknown  30.000000  San Francisco
2  Charlie  18.333333    Los Angeles


In [73]:
df_filled["Age"] = df_filled["Age"].replace(18.3333, round(df_filled["Age"].mean()))
df_filled["Age"] = df_filled["Age"].astype(int)
print(df_filled)

      Name  Age           City
0    Alice   25        Unknown
1  Unknown   30  San Francisco
2  Charlie   18    Los Angeles


# Task 3 Interpretation
I made a table where each column has one missing value (`NaN`).
- `isnull().sum()` counts the missing values in each column.
- `fillna()` replaces them: I used **"Unknown"** for the text columns (Name and City), because filling a name or city with 0 would not make sense.
- For the numeric **Age** column, I first filled the missing value, then replaced it with the **mean (average) age**, which is a more realistic value than 0.

#Rounding the Age
- I used `round()` to turn the mean (e.g. 18.33) into a clean whole number, then `astype(int)` to remove the `.0` decimal.
- Note: `astype(int)` on its own does **not** round — it chops off the decimal (18.99 would become 18, not 19). That is why I round **first**, then convert the type.

# Task 4: GroupBy and Describe

In [74]:
# Using the following DataFrame, group by 'Category' and describe the 'Value' column

df_group = pd.DataFrame({
    'Category': ['A', 'B', 'A', 'B', 'A', 'C'],
    'Value': [10, 20, 15, 25, 30, 35]
})
print(df_group)

  Category  Value
0        A     10
1        B     20
2        A     15
3        B     25
4        A     30
5        C     35


# Task 5: Concatenation and Merging

In [75]:
df1 = pd.DataFrame({'A': [1, 2], 'B': [3, 4]})
df2 = pd.DataFrame({'A': [5, 6], 'B': [7, 8]})
df3 = pd.DataFrame({'C': [9, 10], 'D': [11, 12]})
print("\ndf1")
print(df1)
print("\ndf2")
print(df2)
print("\ndf3")
print(df3)


df1
   A  B
0  1  3
1  2  4

df2
   A  B
0  5  7
1  6  8

df3
    C   D
0   9  11
1  10  12


In [76]:
# Concatenate two DataFrames vertically and merge them horizontally
# Concatenate df1 and df2 VERTICALLY (stack on top of each other)
df_vertical = pd.concat([df1, df2], axis=0, ignore_index=True)
print("After vertical concatenation (df1 on top of df2):")
print(df_vertical)

After vertical concatenation (df1 on top of df2):
   A  B
0  1  3
1  2  4
2  5  7
3  6  8


In [77]:
#Merge the result with df3 HORIZONTALLY
df_merged = pd.concat([df_vertical, df3], axis=1)
print("\nAfter horizontal merge with df3:")
print(df_merged)


After horizontal merge with df3:
   A  B     C     D
0  1  3   9.0  11.0
1  2  4  10.0  12.0
2  5  7   NaN   NaN
3  6  8   NaN   NaN


In [78]:
print("\nFinal shape (rows, columns):", df_merged.shape)


Final shape (rows, columns): (4, 4)



# Task 6: Tuples and Sets

In [79]:
# Create a tuple of fruits (tuples use round brackets and CANNOT be changed)
fruits = ("apple", "banana", "cherry")
print(fruits)

('apple', 'banana', 'cherry')


In [80]:
# Create a set of numbers (sets use curly brackets and hold no duplicates)
numbers_set = {1, 2, 3, 4, 5}
print(numbers_set)

{1, 2, 3, 4, 5}


In [81]:
# Try to add to the SET — this works, sets are changeable (mutable)
numbers_set.add(6)
print("\nSet after adding 6:", numbers_set)


Set after adding 6: {1, 2, 3, 4, 5, 6}


In [82]:
# Try to add to the TUPLE — this fails, tuples are unchangeable (immutable)
try:
    fruits.add("mango")
except AttributeError as e:
    print("\nCannot add to tuple:", e)


Cannot add to tuple: 'tuple' object has no attribute 'add'


#Task 6 Interpretation
- A **set** is *mutable* (changeable), so `.add()` works and 6 was added.
- A **tuple** is *immutable* (fixed), so it has no `.add()` method and trying to change it causes an error.
- I used a `try/except` block to catch the error instead of letting the program crash.
Key difference: tuples are locked after creation; sets can grow and shrink. Sets also automatically remove duplicates.

# Task 7: Dictionaries

In [83]:
# Create a dictionary — keys are student names, values are their scores
students = {"Alice": 85, "Bob": 72, "Charlie": 90}
print("Original:", students)

Original: {'Alice': 85, 'Bob': 72, 'Charlie': 90}


In [84]:
# Update an existing student's score
students["Bob"] = 78
print("After updating Bob:", students)

After updating Bob: {'Alice': 85, 'Bob': 78, 'Charlie': 90}


In [85]:
# Add a brand new student
students["David"] = 88
print("After adding David:", students)

After adding David: {'Alice': 85, 'Bob': 78, 'Charlie': 90, 'David': 88}


# Task 7 Interpretation
A dictionary stores data as **key : value** pairs (here: name : score).
- To **update**, I assigned a new value to an existing key (`students["Bob"] = 78`).
- To **add**, I assigned a value to a brand-new key (`students["David"] = 88`).
Same syntax for both — if the key exists it updates, if it doesn't it creates.

#Task 8: Functions and Lambda & Task 10: Map, Reduce, and Filter

In [86]:
# Regular function to square a number
def square(n):
    return n * n

In [87]:
# Lambda function — a short one-line way to do the same thing
square_lambda = lambda n: n * n

In [88]:
# Test both with two different inputs
print("Regular function:", square(4), square(7))
print("Lambda function:", square_lambda(4), square_lambda(7))

Regular function: 16 49
Lambda function: 16 49


In [89]:
#Lambda with two inputs:
# Add two numbers
add = lambda a, b: a + b
print(add(3, 7))

# Multiply two numbers
multiply = lambda a, b: a * b
print(multiply(4, 5))

10
20


In [90]:
#Lambda with a condition (returns one of two things):
even_or_odd = lambda x: "even" if x % 2 == 0 else "odd"
print(even_or_odd(3))
print(even_or_odd(4))



odd
even


In [91]:
#Lambda on text:
# Make text uppercase
shout = lambda word: word.upper()
print(shout("I am doctor"))

I AM DOCTOR


In [92]:
# Get the length of a word
length = lambda word: len(word)
print(length("banana"))

6


In [93]:
#lambda in inside map and filter

nums = [1, 2, 3, 4, 5]

# Square every number
print(list(map(lambda x: x * x, nums)))

[1, 4, 9, 16, 25]


In [94]:
# Keep only even numbers
print(list(filter(lambda x: x % 2 == 0, nums)))

[2, 4]


### Task 8 Interpretation
Both functions do the same job — square a number.
- The **regular function** uses `def` and is good for longer logic.
- The **lambda** is a shortcut for simple one-line functions.
Both gave the same results (16 and 49), showing they are equivalent for this task.

**I also tried some other lambda examples** to understand them better — such as doubling a number, adding two numbers, checking even or odd, and using lambdas inside `map()` and `filter()`.

**Takeaway:** lambdas are best for short, simple, one-time functions. For anything longer or reused, a regular `def` function is clearer.

# Task 9: Iterators and Generators

In [95]:
# ITERATOR: a class that produces the first 5 even numbers
class EvenNumbers:
    def __init__(self):
        self.current = 0      # start before the first even number
        self.count = 0        # how many we've given out

    def __iter__(self):
        return self           # the object is its own iterator

    def __next__(self):
        if self.count < 5:    # stop after 5 numbers
            self.current += 2
            self.count += 1
            return self.current
        else:
            raise StopIteration   # signal that we're done

print("Iterator result:")
for num in EvenNumbers():
    print(num, end=" ")

Iterator result:
2 4 6 8 10 

In [96]:
# GENERATOR: a function that does the same thing, much shorter
def even_generator():
    num = 2
    for _ in range(5):
        yield num          # 'yield' hands back one value at a time
        num += 2

print("\n\nGenerator result:")
for num in even_generator():
    print(num, end=" ")



Generator result:
2 4 6 8 10 

# Task 9 Interpretation
Both produce the first 5 even numbers (2 4 6 8 10).
- The **iterator** is a class using `__iter__` and `__next__`. It works but takes more code.
- The **generator** is a function using `yield`. It does the same thing in far fewer lines.
`yield` gives back one value each time it's asked, which makes generators a simpler way to build iterators.

# Task 11: Object-Oriented Programming - Creating a Class

In [97]:
# A class is a blueprint for creating objects
class Rectangle:
    # __init__ runs when a new Rectangle is created; it sets the attributes
    def __init__(self, length, width):
        self.length = length
        self.width = width

    # Method to calculate area
    def area(self):
        return self.length * self.width

    # Method to calculate perimeter
    def perimeter(self):
        return 2 * (self.length + self.width)

# Create two Rectangle objects
rect1 = Rectangle(5, 3)
rect2 = Rectangle(10, 4)

print("Rectangle 1 → Area:", rect1.area(), "Perimeter:", rect1.perimeter())
print("Rectangle 2 → Area:", rect2.area(), "Perimeter:", rect2.perimeter())

Rectangle 1 → Area: 15 Perimeter: 16
Rectangle 2 → Area: 40 Perimeter: 28


### Task 11 Interpretation
- The **Rectangle** class is a blueprint. `length` and `width` are its **attributes** (data), and `area()` and `perimeter()` are its **methods** (actions).
- `__init__` sets up each object's attributes when it's created.
- I made two objects with different sizes, and each used the same methods to give its own results — this shows how one class can create many objects.


# Task 12: Pandas Data Analysis

In [98]:
df_employees = pd.DataFrame({
    'Name': ['John', 'Jane', 'Bob', 'Alice', 'Charlie'],
    'Department': ['IT', 'HR', 'IT', 'Finance', 'HR'],
    'Salary': [55000, 65000, 70000, 60000, 58000]
})
print(df_employees)

      Name Department  Salary
0     John         IT   55000
1     Jane         HR   65000
2      Bob         IT   70000
3    Alice    Finance   60000
4  Charlie         HR   58000


In [99]:
# 1. Average salary by department —
# groupby() groups rows by Department,
# we take the mean of Salary for each group
average_salary_by_dept = df_employees.groupby('Department')['Salary'].mean()
print(average_salary_by_dept)

Department
Finance    60000.0
HR         61500.0
IT         62500.0
Name: Salary, dtype: float64


In [100]:
#Get the names of employees with salary > 60000
high_salary_employees = df_employees[df_employees['Salary'] > 60000][['Name', 'Salary']]
print(high_salary_employees)

   Name  Salary
1  Jane   65000
2   Bob   70000


In [101]:
#Add a new column 'Bonus' which is 10% of salary
df_employees['Bonus'] = df_employees['Salary'] * 0.10
df_employees['Bonus'] = df_employees['Bonus'].astype(int)
print(df_employees)

      Name Department  Salary  Bonus
0     John         IT   55000   5500
1     Jane         HR   65000   6500
2      Bob         IT   70000   7000
3    Alice    Finance   60000   6000
4  Charlie         HR   58000   5800


In [102]:
#update salary after adding bonus
df_employees['Salary_updated'] = df_employees['Salary'] + df_employees['Bonus']
print(df_employees)

      Name Department  Salary  Bonus  Salary_updated
0     John         IT   55000   5500           60500
1     Jane         HR   65000   6500           71500
2      Bob         IT   70000   7000           77000
3    Alice    Finance   60000   6000           66000
4  Charlie         HR   58000   5800           63800


### Task 12 Interpretation (Bonus and Updated Salary)
- **Bonus column:** I created a new column by multiplying Salary by 0.10 (10%). Pandas applied this to every row automatically.
- **Salary_updated column:** I then added Salary + Bonus together to show each employee's total pay after the bonus.
- This gives two new columns in the output: the bonus amount on its own, and the final salary including the bonus — so it's easy to see both the extra and the new total side by side.